src/main/java/hudson/security/UserAttributesHelper.java

```diff
@@ -80,90 +84,118 @@
     private static final int ADS_UF_PASSWORD_EXPIRED = 0x80_0000;
 
     // https://ldapwiki.com/wiki/Administratively%20Disabled
-    public static boolean checkIfUserEnabled(@NonNull Attributes user) {
+    public static void checkIfUserEnabled(@NonNull Attributes user) throws DisabledException {
         // Active Directory attributes
         Integer uac = getUserAccountControl(user);
         if (uac != null && (uac & ADS_UF_DISABLED) == ADS_UF_DISABLED) {
-            return false;
+            throw new DisabledException("TODO details");
```

Comment:
Besides avoiding the need to patch other components, a minor benefit is that we can specify diagnostic details right here in the stack trace rather than trying to synchronize log messages.